### PR TITLE
[Images] Support Python 3.9

### PR DESCRIPTION
**Pull request by AlonMaor14**
⚠️ ⚠️ **Moving to Python 3.9**

**Main reasoning and benefits:**
1. Python 3.7 end of support is approaching (2023-06-27).
2. Performance improvements.
3. [PEP 574](https://peps.python.org/pep-0574/), Pickle protocol 5 with out-of-band data. The new type `pickle.PickleBuffer` is rejected when the protocol is smaller than 5.
4. [BPO 1635741](https://bugs.python.org/issue1635741), when Python is initialized multiple times in the same process, it does not leak memory anymore.
5. [BPO 38379](https://bugs.python.org/issue38379), garbage collection does not block on resurrected objects.
6. A number of Python builtins (range, tuple, set, frozenset, list, dict) are now sped up using [PEP 590](https://www.python.org/dev/peps/pep-0590) vectorcall.

**Some useful knowledge:**
1. [PEP 572](https://www.python.org/dev/peps/pep-0572/), Assignment expressions.
2. [PEP 570](https://www.python.org/dev/peps/pep-0570/), Positional-only arguments.
3. [PEP 584](https://www.python.org/dev/peps/pep-0584/), Union Operators in dict.
4. [PEP 616](https://www.python.org/dev/peps/pep-0616/), String methods to remove prefixes and suffixes.

**Makefile changes:**
when using `make <image>`, it's now possible to set `MLRUN_PYTHON_VERSION` to newer versions for example `3.9.13`. When setting `INCLUDE_PYTHON_VERSION_SUFFIX=true`, the image name will have the python major+minor version appended. For example for mlrun-api -> `mlrun-api-py3.9:1.3.0`.

**Notes**:

1. For some images, you will need a base image to be available which depends on the python version. For example `quay.io/mlrun/python:${MLRUN_PYTHON_VERSION}-slim` for mlrun-api.
2. ~~`models-gpu-py39` is introduced separately until it is fully tested. Use `make models-gpu-py39` to build it.~~
3. ~~The actual `CUDA_VER` used in models-gpu py 3.7 is 11.0 (it was being overridden by the makefile) hence the "downgrade"~~
4. ~~Changed the miniconda installer binary in models-gpu (py 3.7) image because latest was giving a python 3.9 installation and we are not there yet (seems like we didn't discover this by now due to the cached layer)~~
5. AST changes - https://github.com/python/cpython/issues/79003
6. Check the new "Migrating to Python 3.9" section in [CONTRIBUTING.md](https://github.com/mlrun/mlrun/blob/development/CONTRIBUTING.md)

### CODE REVIEW

File: dockerfiles/jupyter/Dockerfile [hunk 1]
```diff
@@ -12,7 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-FROM jupyter/scipy-notebook:python-3.8.8
+
+ARG MLRUN_PYTHON_VERSION=3.9.13
+
+FROM jupyter/scipy-notebook:python-${MLRUN_PYTHON_VERSION}
```
Comment (AlonMaor14) [inline dockerfiles/jupyter/Dockerfile:18]: need to upload this image to quay.io/mlrun

File: dockerfiles/models-gpu/py39/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ARG MLRUN_PYTHON_VERSION=3.9.13
+
+ENV CONDA_OVERRIDE_CUDA ${CUDA_VER}
+
+RUN conda config --add channels conda-forge && \
+    conda update -n base -c defaults conda && \
+    conda install -n base \
+        python=${MLRUN_PYTHON_VERSION} \
+        pip~=${MLRUN_PIP_VERSION} \
+        mamba \
```
Comment (AlonMaor14) [inline dockerfiles/models-gpu/py39/Dockerfile:56]: mamba is for debugging and will be rmoved

File: dockerfiles/models/Dockerfile [hunk 3]
```diff
@@ -34,14 +38,15 @@ ENV OMPI_ALLOW_RUN_AS_ROOT=1
...
-ARG TENSORFLOW_VERSION=2.4.1
-# Not using intel-tensorflow cause they have only 2.4.0 and we need 2.4.1
-RUN python -m pip install -U tensorflow~=${TENSORFLOW_VERSION} mxnet
+ARG TENSORFLOW_VERSION=2.9.0
+
+# intel-tensorflow is a slimmer, CPU optimizied distribution of tensorflow by intel
+RUN python -m pip install -U intel-tensorflow~=${TENSORFLOW_VERSION} mxnet
 
-ARG HOROVOD_VERSION=0.22.1
+ARG HOROVOD_VERSION=0.26.1
```
Comment (AlonMaor14) [inline dockerfiles/models/Dockerfile:49]: need to change the versions to match the models gpu py 3.9 image

File: dockerfiles/models-gpu/py39/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+ARG CUDA_VER=11.7.0
+
+#FROM quay.io/mlrun/cuda:${CUDA_VER}-cudnn8-devel-ubuntu18.04
+FROM nvidia/cuda:${CUDA_VER}-base-ubuntu20.04
```
Comment (AlonMaor14) [inline dockerfiles/models-gpu/py39/Dockerfile:17]: needs to upload image to quay

File: dev-requirements.txt [hunk 5]
```diff
@@ -9,13 +9,13 @@ pytest-httpserver~=1.0
...
-matplotlib~=3.0
-graphviz~=0.16.0
+matplotlib~=3.5
+graphviz~=0.20.0
 nuclio-sdk~=0.3.0
 isort~=5.7
 avro~=1.11
 # needed for mlutils tests
-scikit-learn~=1.0
+scikit-learn~=1.0, <1.2
```
Comment (AlonMaor14) [inline dev-requirements.txt:18]: scikit learn 1.2 has breaking changes

**Review by liranbg — APPROVED** [review]

**Pull request merged**